ern/index.go

```diff
@@ -123,70 +123,75 @@ func isUniqueErr(err error) bool {
-// MessageThreadId, MessageSender, MessageRecipient and MessageCreatedDateTime.
-func (i *Indexer) indexMessageHeader(ernID *cid.Cid, obj *meta.Object) error {
-	graph := meta.NewGraph(i.store, obj)
+ // decode decodes whatever is stored at path into the given value
+func DecodeIndexObj(i *Indexer, metaObj *meta.Object, v interface{}, path ...string) (err error) {
+	graph := meta.NewGraph(i.store, metaObj)
 
-	// decode decodes whatever is stored at path into the given value
-	decode := func(v interface{}, path ...string) (err error) {
-		defer func() {
-			if err != nil {
-				err = fmt.Errorf("error decoding %s into %T: %s", path, v, err)
-			}
-		}()
-		x, err := graph.Get(path...)
-		if meta.IsPathNotFound(err) {
-			return nil
-		} else if err != nil {
-			return err
-		}
-		id, ok := x.(*cid.Cid)
-		if !ok {
-			return fmt.Errorf("expected %s to be *cid.Cid, got %T", path, x)
-		}
-		obj, err := i.store.Get(id)
+	defer func() {
 		if err != nil {
-			return err
+			err = fmt.Errorf("Error decoding %s into %T: %s", path, v, err)
 		}
-		return obj.Decode(v)
+	}()
+
+	x, err := graph.Get(path...)
+	if meta.IsPathNotFound(err) {
+		return err
+	} else if err != nil {
+		return err
+	}
+	id, ok := x.(*cid.Cid)
+	if !ok {
+		return fmt.Errorf("Expected %s to be *cid.Cid, got %T", path, x)
 	}
 
-	// insert the MessageSender and MessageRecipient into the party index
-	insertParty := func(field string) (*cid.Cid, error) {
-		var id *cid.Cid
-		link, err := obj.GetLink(field)
-		if err == nil {
-			id = link.Cid
-		} else if !meta.IsPathNotFound(err) {
-			return nil, err
-		}
-		var partyID struct {
-			Value string `json:"@value"`
-		}
-		if err := decode(&partyID, field, "PartyId"); err != nil {
-			return nil, err
-		}
-		var partyName struct {
-			Value string `json:"@value"`
-		}
-		if err := decode(&partyName, field, "PartyName", "FullName"); err != nil {
-			return nil, err
-		}
-		_, err = i.db.Exec(
-			"INSERT INTO party (cid, id, name) VALUES ($1, $2, $3)",
-			id.String(), partyID.Value, partyName.Value,
-		)
-		if err != nil && !isUniqueErr(err) {
-			return nil, err
-		}
-		return id, nil
+	obj, err := i.store.Get(id)
+	if err != nil {
+		return err
+	}
+	return obj.Decode(v)
+}
+
+// insert the MessageSender and MessageRecipient into the party index
+func InsertParty(i *Indexer, metaObj *meta.Object, field string) (*cid.Cid, error) {
+	var id *cid.Cid
+	link, err := metaObj.GetLink(field)
+	if err == nil {
+		id = link.Cid
+	} else if !meta.IsPathNotFound(err) {
+		return nil, err
+	}
+	var partyID struct {
+		Value string `json:"@value"`
 	}
-	sender, err := insertParty("MessageSender")
+	if err := DecodeIndexObj(i, metaObj, &partyID, field, "PartyId"); err != nil {
+		pid := &partyID
+		pid.Value = "000"
+	}
+	var partyName struct {
+		Value string `json:"@value"`
+	}
+	if err := DecodeIndexObj(i, metaObj, &partyName, field, "PartyName", "FullName"); err != nil {
+		return nil, err
+	}
+	_, err = i.db.Exec(
```

Comment:
Wrap all sql transaction with Begin and Commit.
See https://github.com/meta-network/go-meta/blob/master/cwr/index.go#L76 for reference .